### PR TITLE
Add a `copysign` function to f32 and f64

### PR DESCRIPTION
**Pull request by raphlinus**
This patch adds a `copysign` function to the float primitive types. It is an exceptionally useful function for writing efficient numeric code, as it often avoids branches, is auto-vectorizable, and there are efficient intrinsics for most platforms.

I think this might work as-is, as the relevant `copysign` intrinsic is already used internally for the implementation of `signum`. It's possible that an implementation might be needed in japaric/libm for portability across all platforms, in which case I'll do that also.

Part of the work towards #55107

### CODE REVIEW

**Comment by joshtriplett** [comment]
The two versions have doc comments that explain the `NAN` behavior differently. Could you please make them match? I found the explanation on the `f32` version much clearer; could you copy that to the `f64` version?

**Comment by raphlinus** [comment]
Definitely, and good catch. Was just a copy-paste error.

Who is a good person to look into the question of whether this might break exotic targets because llvm might generate a call to a (possibly unimplemented) library function for the general copysign intrinsic. Maybe @japaric?

**Comment by nikic** [comment]
Instead of `a.copysign(b)` something like `a.with_sign_of(b)` might make it more explicit which of the signs is being applied to which of the operands. On the other hand, `copysign` is standard terminology.

**Comment by raphlinus** [comment]
@nikic Good point, and I'm open to that. In this case, I'm going to suggest sticking with standard terminology, because I think the target user is much more likely to be studying or adapting numerical algorithms from other languages, rather than than writing Rust-only numerical algorithms. Also I think when studying asm output (especially wasm), it reduces the translation, understanding that `copysign` in the source should translate into the `copysign` intrinsic.

But if people strongly prefer a more explicit name, I'll change it.

**Comment by joshtriplett** [comment]
I think keeping the standard name `copysign` makes sense. And you've kept the order of arguments: `a.copysign(b)` matches `copysign(a, b)`.

You *could* make it `copysign_of` for clarity, but even that divergence from the standard function might do more harm than good. I'll leave that up to you.

However, that brings to mind another thought. To prevent someone from thinking that this might modify the value in place, please add `#[must_use]` to require using the return value.

**Comment by raphlinus** [comment]
@joshtriplett Done. I agree with the rationale but will point out it's not consistent with the rest of the module. I think we can chalk that up to `must_use` not being available when the base library was done.

**Comment by joshtriplett** [comment]
LGTM. If you would like to seek feedback to confirm that this will work on obscure architectures, please feel free to do so and r=me afterward. You could also go ahead, on the theory that we already use many other intrinsics and this one doesn't seem drastically *less* likely to work.

**Comment by raphlinus** [comment]
I'm comfortable going ahead and seeing what breaks. If we get a failure on a specific architecture because of a missing `_copysign`, it should be pretty clear what to do next to fix it.

File: src/libstd/f32.rs [hunk 1]
```diff
@@ -198,6 +198,35 @@ impl f32 {
         }
     }
 
+    /// Returns a number composed of the magnitude of one number and the sign of
+    /// another.
+    ///
+    /// Equal to `self` if the sign of `self` and `y` are the same, otherwise
+    /// equal to `-y`. If `self` is a `NAN`, then a `NAN` with the sign of `y`
```
Comment (LukasKalbertodt) [inline src/libstd/f32.rs]: "otherwise equal to `-self`", and not "`-y`", right? Or am I confused now?

Also: I'd rephrase the first sentence to "Returns a number composed of the magnitude of `self` and the sign of `y`". That's more explicit. 

And yes, I know the PR is already merged, but I guess we should fix this in a new PR, if I'm right about this typo.

Comment (raphlinus) [reply src/libstd/f32.rs]: @LukasKalbertodt You're absolutely right. I'm happy to fix this in a followup PR. Good catch!

**Comment by raphlinus** [comment]
@crlf0710 I was told on IRC that this issue could serve as the tracking issue, for such a small feature. But I am not expert on the ways of Rust stabilization.

**Comment by crlf0710** [comment]
@raphlinus The tracking issue should be an issue, not a PR, i think... @Centril

**Comment by raphlinus** [comment]
Ok, will do. One reason I've been holding back is that was also planning on adding "round to nearest even" function as part of the same motivating work, but now I see that could take a long time (honestly, the motivation I have for it has somewhat stalled out), so no reason that should block this.

**Comment by crlf0710** [comment]
Well i've just finished seeking for solutions of "round with ties to positive infinity" when i come to this issue yesterday. :)
